physics/rrtmgp_sw_cloud_optics.F90

```diff
@@ -111,159 +108,269 @@ subroutine rrtmgp_sw_cloud_optics_init(doG_cldoptics, doGP_cldoptics_PADE, doGP_
+    call mpi_bcast(nSize_liqSW,        1, MPI_INTEGER, mpiroot, mpicomm, mpierr)
+    call mpi_bcast(nSize_iceSW,        1, MPI_INTEGER, mpiroot, mpicomm, mpierr)
+    call mpi_bcast(nSizeregSW,         1, MPI_INTEGER, mpiroot, mpicomm, mpierr)
+    call mpi_bcast(nCoeff_extSW,       1, MPI_INTEGER, mpiroot, mpicomm, mpierr)
+    call mpi_bcast(nCoeff_ssa_gSW,     1, MPI_INTEGER, mpiroot, mpicomm, mpierr)
+    call mpi_bcast(nBoundSW,           1, MPI_INTEGER, mpiroot, mpicomm, mpierr)
+    call mpi_bcast(nPairsSW,           1, MPI_INTEGER, mpiroot, mpicomm, mpierr)
+#endif
+    
+    ! Has the number of ice-roughnesses provided from the namelist?
+    ! If so, override nrghice from cloud-optics file
+    if (nrghice .ne. 0) nrghice_fromfileSW = nrghice
+#ifdef MPI
+    call mpi_bcast(nrghice_fromfileSW, 1, MPI_INTEGER, mpiroot, mpicomm, mpierr)
+#endif
+
+    ! #######################################################################################
+    !
+    ! Allocate space for arrays...
+    ! (ALL processors)
+    !
+    ! #######################################################################################
+    if (doGP_cldoptics_LUT) then
+       allocate(lut_extliqSW(nSize_liqSW, nBandSW))
+       allocate(lut_ssaliqSW(nSize_liqSW, nBandSW))
+       allocate(lut_asyliqSW(nSize_liqSW, nBandSW))
+       allocate(lut_exticeSW(nSize_iceSW, nBandSW, nrghice_fromfileSW))
+       allocate(lut_ssaiceSW(nSize_iceSW, nBandSW, nrghice_fromfileSW))
+       allocate(lut_asyiceSW(nSize_iceSW, nBandSW, nrghice_fromfileSW))
+    endif
+    if (doGP_cldoptics_PADE) then
+       allocate(pade_extliqSW(nBandSW, nSizeRegSW,  nCoeff_extSW ))
+       allocate(pade_ssaliqSW(nBandSW, nSizeRegSW,  nCoeff_ssa_gSW))
+       allocate(pade_asyliqSW(nBandSW, nSizeRegSW,  nCoeff_ssa_gSW))
+       allocate(pade_exticeSW(nBandSW, nSizeRegSW,  nCoeff_extSW,   nrghice_fromfileSW))
+       allocate(pade_ssaiceSW(nBandSW, nSizeRegSW,  nCoeff_ssa_gSW, nrghice_fromfileSW))
+       allocate(pade_asyiceSW(nBandSW, nSizeRegSW,  nCoeff_ssa_gSW, nrghice_fromfileSW))
+       allocate(pade_sizereg_extliqSW(nBoundSW))
+       allocate(pade_sizereg_ssaliqSW(nBoundSW))
+       allocate(pade_sizereg_asyliqSW(nBoundSW))
+       allocate(pade_sizereg_exticeSW(nBoundSW))
+       allocate(pade_sizereg_ssaiceSW(nBoundSW))
+       allocate(pade_sizereg_asyiceSW(nBoundSW))
+    endif
+    allocate(band_limsCLDSW(2,nBandSW))
 
-       ! Read in fields from file
+    ! #######################################################################################
+    !
+    ! Read in data ...
+    ! (ONLY master processor(0), if MPI enabled) 
+    !
+    ! #######################################################################################
+#ifdef MPI
+    if (mpirank .eq. mpiroot) then
+#endif 
        if (doGP_cldoptics_LUT) then
           write (*,*) 'Reading RRTMGP shortwave cloud data (LUT) ... '
           status = nf90_inq_varid(ncid,'radliq_lwr',varID)
```

Comment:
The code should check whether these call succeeded, and at least report that they failed. Otherwise, when a failure occurs, you'll get bad results without knowing why.